twilio_integration/overrides/notification.py

```diff
@@ -1,38 +1,127 @@
+				frappe.throw(_("Twilio WhatsApp Number is required in Twilio Settings."))
+
+	def send(self, doc, context=None):
+
+		if not context:
+			context = get_context(doc)
+
+		context.update({"doc": doc, "alert": self, "comments": None})
 
-	def send(self, doc):
-		context = get_context(doc)
-		context = {"doc": doc, "alert": self, "comments": None}
 		if doc.get("_comments"):
 			context["comments"] = json.loads(doc.get("_comments"))
 
 		if self.is_standard:
 			self.load_standard_properties(context)
 
 		try:
-			if self.channel == 'WhatsApp':
+			if self.channel == 'WhatsApp' and self.enabled:
 				self.send_whatsapp_msg(doc, context)
-		except:
-			frappe.log_error(title='Failed to send notification', message=frappe.get_traceback())
+		except Exception as e:
+			frappe.log_error(
+				message=frappe.get_traceback(),
+				title=_("Failed to send WhatsApp Notification: {0} for {1} {2}").format(self.name, doc.doctype, doc.name)
+			)
 
-		super(SendNotification, self).send(doc)
+		super(SendNotification, self).send(doc,context=context)
 
 	def send_whatsapp_msg(self, doc, context):
+		notification_type = self.get_notification_type()
+		receiver_list = self.get_receiver_list(doc, context)
+
+		if not receiver_list:
+			return
+
+		formatted_receiver_list = self.format_numbers_for_whatsapp(receiver_list)
+
+		if not formatted_receiver_list:
+			return
+
+		message_content = frappe.render_template(self.message, context)
+
+		ref_doctype = get_reference_doctype(doc)
+		ref_name = get_reference_name(doc)
+
+		self.create_communication(doc, message=message_content, receiver_list= formatted_receiver_list)
+
+		if notification_type:
+			set_notification_last_scheduled(ref_doctype, ref_name, notification_type, "WhatsApp")
+
+
 		WhatsAppMessage.send_whatsapp_message(
-			receiver_list=self.get_receiver_list(doc, context),
-			message=frappe.render_template(self.message, context),
-			doctype = self.doctype,
-			docname = self.name
-		)
\ No newline at end of file
+			receiver_list=formatted_receiver_list,
+			message=message_content,
+			doctype = ref_doctype,
+			docname = ref_name,
+			notification_type=notification_type
+		)
+
+	def create_communication(self, doc, message=None, receiver_list=None):
+		try:
+			timeline_doctype, timeline_name = self.get_timeline_doctype_and_name(doc)
+
+			communication = frappe.get_doc({
+				"doctype": "Communication",
+				"communication_type": "Automated Message",
+				"communication_medium": "Other",
+				"subject": f"WhatsApp: {doc.name}",
+				"content": message,
+				"sent_or_received": "Sent",
+				"reference_doctype": get_reference_doctype(doc),
+				"reference_name": get_reference_name(doc),
+				"sender": frappe.session.user,
+				"recipients": "\n".join(receiver_list),
+				"phone_no": receiver_list[0] if len(receiver_list) == 1 else None
+			})
+
+			communication.append("timeline_links", {
+				"link_doctype": timeline_doctype,
+				"link_name": timeline_name
+			})
```

Comment:
this part should be conditional, append only if `timeline_doctype` and `timeline_name` is truthy